### ISSUE TITLE
Should all this event state switching be automated?

### ISSUE THREAD

**Issue by ttencate** [alakajam-team/alakajam#386]
It's not nice that someone has to press a button at a specific time.

The naive solution would be to create a cron job (or cron-like system in Node), but it's an extra piece of machinery, hard to test, and somewhat brittle.

So instead, I'm thinking of a more stateless data model where the `Event` object itself only contains time-independent properties like `title` and `logo`. All the time-dependent properties would go into a separate object, say `EventState`, which has `(event_id, start_time)` as its primary key. For a given `event_id`, at a given moment, the `EventState` with the most recent `start_time` that's not in the future is considered the "current" one. It's one extra join on indexed columns of very small tables, so the performance impact should be negligible.

In the event admin UI, there could be tabs along the left or top, listing the start times. These can be clicked to select that particular state for editing. There are also controls for adding and removing times from the list. And the current one is clearly highlighted. I can make a mockup if it's not clear.

Scheduled blog posts could maybe somehow be tied into it as well.

I'm just throwing this idea out there so we can discuss it; I don't have time to build it any time soon :( But I think it would make the organization of events more robust, because anyone can set things up beforehand, and the thing then basically runs itself.

**Comment by Aurel300** [comment]
I mostly agree. I think it's an overall more robust design – the only good thing about the manual mode is that an admin *has* to see what is happening, i.e. it is a last line of defence if something goes horribly wrong ("and the winning theme is … `undefined`!"). But admittedly these sorts of things should be caught and fixed a long time before they happen.

Additionally these timed sub-events ("theme selected", "jam started", "jam finished", etc) could be integrated into some sort of a template, so when a new Alakajam is scheduled, the admin only chooses one time, e.g. the main jam start, and the sub-events are scheduled relative to that. This is better than having to follow a Google doc script for sure.

**Comment by mkalam-alami** [comment]
I agree with all this, automated jams have been dreamed of since early site development :D The good news is there are some recent features that should help with that:

* "[Event presets](https://github.com/alakajam-team/alakajam/blob/master/core/models.js#L285)" *(see screenshot at the bottom)* clone all(?) of those time-dependant properties, with the goal to simplify event administration. The idea is that instead of flipping all those switches and links and messages, we'd just have to choose a preset from a dropdown menu - then optionally customize the fields we want. The feature is not thoroughly tested so I haven't used it much for now.
* "[Event templates](https://github.com/alakajam-team/alakajam/blob/master/core/models.js#L328)" allow to quickly create an event with pre-filled data.

My original plan was to build upon them for automation. The next layer as I pictured it would be a cron system able to flip the presets at the right time, with possibly a way to integrate blog posts & more. The "event templates" would probably be used to hold the sequencing of the presets.

The idea of joining on a `start_time` column instead of having a full cron system is tempting, as it simplifies things a lot. My only concern is that some switches are supposed to trigger data processing (eg. computing the theme shortlist, or jam rankings), which means we'd need a way to detect when a new state is activated for the first time, and resolve the work synchronously. I guess it's doable as long as the workload is light (a few seconds tops) and we deal with concurrent access to the site.

On the other hand, I feel a cron system might become important if the site grows, as recalculating things on the fly like we do know (like Karma and game ratings) may become a burden in the long run. Oh and cron tasks are also seriously considered for email notifications (#375).

![alakajam com_admin_event-presets_edit 3](https://user-images.githubusercontent.com/694508/52914801-efd5e380-32cc-11e9-957f-0bdbafcdd4f7.png)
